README.md

```diff
@@ -1,2 +1,313 @@
+        votedFor := cId
+        send (VoteResponse, nodeld, currentTerm, true) to node cId
+    else
+        send (VoteResponse, nodeld, currentTerm, false) to node cId
+    end if
+end on
+```
+This phase handles the voting process when a candidate requests votes from other nodes.
+
+* If the received term is greater than the current term, the node updates its term and role.
+* The candidate's log is checked against the follower's log.
+* If the follower has not voted in this term, and the candidate's log is at least as up-to-date as the follower's log, the follower votes for the candidate. Otherwise, it denies the vote.
+
+## Collecting Votes
+
+```shell
+on receiving (voteResponse, voterId, term, granted) at nodeld do
+    if currentRole = candidate and term = currentTerm and granted then
+        votesReceived := votesReceived U {voterId}
+        if votesReceived > (|nodes| + 1) / 2 then
+            currentRole := leader
+            currentLeader := nodeld
+            cancel election timer
+            for each follower in nodes \ {nodeld} do
+                sentLength[follower] := log.length
+                ackedLength[follower] := 0
+                REPLICATELOG(nodeld, follower)
+            end for
+        end if
+    else if term > currentTerm then
+        currentTerm := term
+        currentRole := follower
+        votedFor := null
+        cancel election timer
+    end if
+end on
+
+```
+
+This phase handles the response to a vote request from a candidate.
+
+* If the node is a candidate and receives a vote response for the current term, it adds the vote to the received votes.
+* If the candidate receives votes from a majority of the nodes, it becomes the leader.
+* If the received term is greater than the current term, the node updates its term and role.
+
+## Broadcasting Messages
+```shell
+on request to broadcast msg at node nodeld do
+    if currentRole = leader then
+        append the record (msg: msg, term: currentTerm) to log
+        ackedLength[nodeld] := log.length
+        for each follower in nodes \ {nodeld} do
+            REPLICATELOG(nodeld, follower)
+        end for
+    else
+        forward the request to currentLeader via a FIFO link
+    end if
+end on
+
+periodically at node nodeld do
+    if currentRole = leader then
+        for each follower in nodes \ {nodeld} do
+            REPLICATELOG(nodeld, follower)
+        end for
+    end if
+end on
+
+```
+
+This phase handles broadcasting messages in the cluster.
+
+* If the node is the leader, it appends the message to its log and replicates it to the followers.
+* If the node is not the leader, it forwards the request to the current leader.
+
+## Replicating from Leader to Followers
+
+```shell
+function REPLICATELOG(leaderId, followerId)
+    prefixLen := sentLength[followerId]
+    suffix := (log[prefixLen+1:], ...)
+    prefixTerm := 0
+    if prefixLen > 0 then
+        prefixTerm := log[prefixLen].term
```

Comment:
I think it should be "prefixTerm := log[prefixLen - 1].term" instead of 'prefixLen''th index.